### DISCUSSION TITLE
watch(callback) doesn't trigger when enabling an input field

### DISCUSSION THREAD

**Discussion by hanskr** [react-hook-form/react-hook-form#5326] [Q&A]
I have a form with a checkbox that enables/disables another input field. And I use a `watch(cb)` to render a live preview of the results. 

The problem is that when enabling the input field, the value passed to watch from my input field is `undefined`, as the field had not yet been enabled. And there is no retriggering watch once the fileds becomes active. 

Is this a bug, or is there a way to work around this?

**Comment by bluebill1049** [comment]
can you provide a codesandbox to reproduce this?

**Reply by hanskr** [reply]
https://codesandbox.io/s/blissful-panini-zsqry?file=/src/App.tsx

If you look at the console there, and click the checkbox a few times, you can see that the "preview" it would render has values that are from "before" the input field gets disabled. 

As for why I use two different watch/useWatch it is because in my real life form it is huge and nested.

**Reply by bluebill1049** [reply]
> If you look at the console there, and click the checkbox a few times, you can see that the "preview" it would render has values that are from "before" the input field gets disabled.

I couldn't reproduce the issue, what do you mean it's before?

**Reply by hanskr** [reply]
I updated the example to be clearer now. 

When you click the "disable name" checkbox, the preview will still render the name, and when you reenable the name field, the preview will not pick up the name from the newly enabled input element.

**Reply by bluebill1049** [reply]
https://user-images.githubusercontent.com/10513364/119489303-e63a6080-bd9e-11eb-9c6b-50f0c1feeb0b.mov

which part?

**Reply by hanskr** [reply]
Maybe I'm explaining this badly? But when the checkbox is _enbled_ the preview should _not_ show a name, and vice-versa. Which is opposite of what happens right now. (I guess the reason this happens is that the change that is triggered from the click on checkbox is handled before the actual input field is enabled again)

**Reply by bluebill1049** [reply]
oh i see your point now. will take a look at it.

**Reply by bluebill1049** [reply]
This is not a bug. here is flow

-> click
-> render (your console is here)
-> checkbox set true, disabled -> input

so my recommendation is derived that input value with checkbox state, otherwise, it's a chicken and egg situation.

**Reply by hanskr** [reply]
I'm not sure I understand your reccomendation. 

My problem is that I don't see a place to hook into that will allow me to react on the input fields being set to enabled. If I react on the click on the checkbox it is too soon, and there is no onEnabled/Disabled hook. 

So I'm at a loss as to when to trigger the rerendering of my preview to be able to pick up the actual current value. 

Only thing I got working so far is setting a timeout on the watch callback and using `getValues()`, but that feels super dirty. 

```
const cb = (value) => {
    setTimeout(() => {
      setPreview(getValues());
    }, 100);
};
```

**Reply by bluebill1049** [reply]
because setting input disabled it's always going to be one render behind.

**Reply by hanskr** [reply]
Yes, that is my point as well. But I'm wondering if there is a nice workaround for this. I think it is a very legitimate use-case to a) disable/enable fields in a form and b) rendering a preview based on the current contents of the form. 

Could perhaps react-hook-form be conficured to trigger on a enable/disable of fields that are registered?

**Reply by hanskr** [reply]
FWIW I found a slightly less hackish workaround for this now: 

```
    const [forceRender, setForceRender] = useState(false);

    function deferPreviewUpdate() {
        setForceRender((prevValue) => !prevValue);
    }

    useEffect(() => {
        updatePreview(getValues());
    }, [forceRender]);

    const cb: WatchObserver<BasicProfileOrgMotor> = (value, { name }) => {
            deferPreviewUpdate();
    };

    useEffect(() => {
        watch(cb);
    }, []);

```
